### ISSUE TITLE
[CPU] logit_backward and six activation backward kernels return a different gradient at NaN depending on tensor length

### ISSUE THREAD

**Issue by qflen** [pytorch/pytorch#195075]
### Describe the bug

Seven CPU backward kernels return a different gradient at NaN depending on how long the tensor is. `cpu_kernel_vec` runs the vectorized lambda over full SIMD blocks and the scalar lambda over the remainder, and in each of these the vector lambda is written as the ordered negation of the scalar predicate. That rewrite is only valid for ordered inputs, so at NaN the two take opposite branches.

The clearest form: one tensor, nine identical NaN elements, two different gradients inside it.

```python
import torch
import torch.nn.functional as F

def grad9(fn):
    x = torch.full((9,), float("nan"), requires_grad=True)
    fn(x).sum().backward()
    return x.grad[0].item(), x.grad[8].item()

print("logit_backward eps=0.1", grad9(lambda x: torch.special.logit(x, eps=0.1)))
for name, fn in [("hardtanh", F.hardtanh), ("hardshrink", F.hardshrink), ("softshrink", F.softshrink),
                 ("elu", F.elu), ("celu", F.celu), ("hardswish", F.hardswish)]:
    print(f"{name:<12}", grad9(fn))
```

float32 on an M5 (arm64 NEON, vector block 8), `grad[0]` then `grad[8]`:

```
logit_backward eps=0.1 (0.0, nan)
hardtanh     (0.0, 1.0)
hardshrink   (0.0, 1.0)
softshrink   (0.0, 1.0)
elu          (nan, 1.0)
celu         (nan, 1.0)
hardswish    (nan, 1.0)
```

`eps=None` is unaffected (that branch falls back to NaN on both sides), and the forwards are unaffected. For hardtanh, hardshrink and softshrink the vector path also swallows a NaN gradient, so `detect_anomaly` fires at length 7 and stays quiet at length 8.

### Instances

| kernel | scalar lambda | vector lambda | scalar at NaN | vector at NaN |
| --- | --- | --- | --- | --- |
| `logit_backward`, `eps >= 0` branch, `cpu/BinaryOpsKernel.cpp:954-969` | `(x < lo \|\| x > hi) ? 0 : dy / (x * (1 - x))` | `blendv(0, dy / (x * (1 - x)), (x >= lo) & (x <= hi))` | `nan` | `0.0` |
| `hardtanh_backward_kernel`, `cpu/Activation.cpp:724` / `:727` | `(self <= min \|\| self >= max) ? 0 : grad` | `((self > min) & (self < max)) & grad` | `grad` | `0` |
| `shrink_backward_kernel` (hardshrink, softshrink), `:689` / `:693` | `(self >= -lambd && self <= lambd) ? 0 : grad` | `((self < -lambd) \| (self > lambd)) & grad` | `grad` | `0` |
| `elu_backward_kernel` (elu, celu), `:266` / `:270` | `b <= 0 ? <neg branch> : a * poscoef` | `blendv(<neg branch>, a * poscoef, b > 0)` | `grad * scale` | `nan` |
| `hardswish_backward_kernel`, `:846-854` / `:855-865` | three-way if/else ending in `return grad` | nested `blendv` | `grad` | `nan` |

Each activation kernel has a reduced-float twin with the same shape (`:705-715`, `:232-248`, `:809-831`). CUDA implements the scalar form in all of these, from reading the source; I do not have a CUDA machine.

### The convention has to be named

The vector lambdas are not simply wrong. The forward-mode AD formulas in `tools/autograd/derivatives.yaml:2534` (hardtanh), `:2181` (hardshrink) and `:2683` (softshrink) already side with them. For `logit_backward` the same split exists: `torch/_decomp/decompositions.py` and `infinitely_differentiable_logit_backward` give `0.0`, while CUDA and the CPU scalar lambda give NaN.

So each op has several implementations that disagree, and a fix needs a maintainer to say whether NaN propagates or is absorbed to 0. The length dependence is a bug under either answer.

### Related

Same class, filed by @krrishapatel: #193753, #193754, #193781, with #77742 as the closed ancestor. Different ops, same failure mode: a scalar lambda and a vector lambda in one `cpu_kernel_vec` that are not equivalent on special values, so the tensor length decides the answer.

### Versions

```
PyTorch version: 2.15.0a0+gitf5f7ac4
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
Clang version: 17.0.0 (clang-1700.4.4.1)
CMake version: version 4.3.1

Python version: 3.13.5 (main, Jun 12 2025, 12:22:43) [Clang 20.1.4 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
Is XPU available: False

CPU: Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] optree==0.19.0
[pip3] torch==2.15.0a0+gitf5f7ac4
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @snadampal @milpuz01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia

**Comment by AnkitNakhawa** [comment]
I reproduced this and opened draft PR #195204 with regression coverage, including the reduced-precision paths. The proposed fix makes the CPU SIMD predicates match the existing scalar/CUDA behavior at NaN. Would maintainers agree with this direction and consider marking the issue actionable?
